### ISSUE TITLE
How to use a social provider SSO for authentication to secure the REST APIs using OAuth

### ISSUE THREAD

**Issue by tuhinkanti** [spring-guides/tut-spring-boot-oauth2#30]
This is a great tutorial on REST and we are able to bring up a pretty functional app using this. I am facing a challenge with OAuth though. In this tutorial we are providing credentials to get authenticated. We want to use github(google) for authentication. I have studied in detail the OAuth and Angular JS examples, they all concern a single page/UI app. If I try to use the bearer token from there in an API request, the authentication is anonymous and it fails. I am unable to figure out, how I could possibly use OAuth for a REST call while getting authentication from a social provider on a browser. I envision:
1) Authenticate using social provider through a UI
2) Grab the bearer token
3) Use the bearer token with additional client identifiers for an API request
4) Get access to the protected resource.
This enables me to not have to deal with user credentials. Could you suggest an approach?
Many thanks and all the documentation thus far has been extremely wonderful.

**Comment by dsyer** [comment]
IMO your angular (or any browser based app) should never see an acces token. Who is it calling your API? Who hosts the API (is it yours or the social provider's)? These questions will be relevant if you are trying to figure out what the access token means for your client app. The access token value is easy to get at in the app (it's stored as part of the "details" in the authentication token), but what you can or should do with it is not cut and dried.

**Comment by tuhinkanti** [comment]
The browser based app is a facade to enable the end user to see the social logic page(the provider it trusts). The APIs are hosted by my application and the tokens I would like to be generated my application, I want to use the social provider for authentication. I saw the token details after downcasting the Authentication object as you mentioned and these are all good from the providers standpoint.
In  a gist, the workflow I though might make sense
1) User access' my one page application(facade) to authenticate.
2) My application gets the authentication from social(authorization from the providers standpoint), generates token for authorization for API's hosted by me
3) User will have access to the token my app generated and can use this for any API calls. Step 3 fails if I understand correctly the EnableSSO workflow, since when I provide the bearer token on a rest client it treats the authentication as anonymous. Any resource accessed from the browser is though correctly authorized

An ideal workflow if possible:
1) User authenticates based on a provider
2) My application authorizes based on the "details" of the provider token
3) I can use this token stateless for any subsequent REST requests

From what I understand about OAuth not many authentication providers(social providers) can secure your credentials and hence they do not expose password flow, refresh token flow. They expose web server flow and authorization code flow in which case a browser is required. I am able to do the above with an LDAP provider or a provider which accepts credentials(just like this tutorial) but in a production application, this looks insecure as I my application inadvertently has access to credentials while it doesn't own the authentication.
Thanks much for taking out time to reply, do let me know if what I said above isn't a correct expectation.

**Comment by dsyer** [comment]
Thanks for the extra detail. I'm not sure "the user who authenticated is calling my APIs" is clear enough yet. How is the user doing this? With his own app? Through a browser? From the command line?

I also think "my application inadvertently gets access to credentials" is the heart of the reason that OAuth2 exists. You want to avoid that right? So you centralize the authentication and token generation. You don't want to "secure the credentials". You just want to delegate all of that to a service (possibly one that you provide, or one of the "social" providers if your needs are simple).

I'm not sure we are making progress yet, really, since I haven't actually seen a well posed question.

**Comment by tuhinkanti** [comment]
Let me try this one last time. I am running social-auth-server[tut-spring-boot-oauth2] tutorial. 
I run the SocialApplication.java, I am asked to authenticate using github(I am not using Facebook here) on the browser. I get the Hello "user" greeting when I browse to http://localhost:8080/user(protected resource). 
I run the ClientApplication.java, open http://localhost:9999/client. From this request, I grab the OAuth token from http://localhost:8080/trace(I cheat since I am just looking for a mechanism).
I do a curl -v http://localhost:9999/client -H "authorization: bearer $TOKEN" on the command line and I see a 302 in the verbose output. In the application logging I see: Access is denied (user is anonymous); redirecting to authentication entry point.
The same happens when I run the oauth2-vanilla tutorial.
In the bookmarks example if I provide credentials for getting the bearer token, I am able to authenticate subsequent requests. 
To summarize, I am unable to find out the reason that if the bearer token is through the browser any request on that browser session is authenticated while any other request: incognito, curl needs re-auth(Is this session based?). In case of bookmarks though the auth is somewhat persistent.
I want to know the way I can persist the auth(like bookmarks) while authenticating through a browser? Apologies for not being clear earlier. 
Thanks much for your attention, and if this isn't clear yet you may resolve this.

**Comment by dsyer** [comment]
I'm happy to continue discussing this, but I don't think we are on the same wavelength yet. You gave some concrete examples of using curl on the command line to attempt to connect using bearer tokens to an app that wasn't configured to accept them. Unsurprisingly, that didn't work. But then you said that you are "unable to find out the reason for the bearer token to work through the browser". It doesn't. The browser doesn't, and shouldn't, know the access token. So the command line example wasn't very helpful. The best way to clarify what you mean is probably to just publish some code. Use a fork of this repository, or your own code if you prefer, and post some detailed steps to achieve your goals, and explain how they don't. Maybe we can make some progress then.

**Comment by tuhinkanti** [comment]
Thanks for your prompt response. I realize that I did not understand the workflow correctly.
1) I first run SocialApplication.java from social-auth-server[tut-spring-boot-oauth2] tutorial.
2) Then I run ClientApplication.java from social-auth-server[tut-spring-boot-oauth2]. I browse to http://127.0.0.1:9999/client/ which redirects me to http://localhost:8080 where I authenticate using github and I get Hello "user" getting redirected back to http://127.0.0.1:9999/client post authentication. Following is the trace:
`[
{
timestamp: 1469507970813,
info: {
method: "GET",
path: "/me",
headers: {
request: {
authorization: "bearer 9c03d406-14ac-4f98-a261-6ad29f81fe6a",
accept: "application/json",
user-agent: "Java/1.8.0_65",
host: "localhost:8080",
connection: "keep-alive"
},
response: {
X-Content-Type-Options: "nosniff",
X-XSS-Protection: "1; mode=block",
Cache-Control: "no-cache, no-store, max-age=0, must-revalidate",
Pragma: "no-cache",
Expires: "0",
X-Frame-Options: "DENY",
X-Application-Context: "application",
Content-Type: "application/json;charset=UTF-8",
Transfer-Encoding: "chunked",
Date: "Tue, 26 Jul 2016 04:39:30 GMT",
status: "200"
}
}
}
},
{
timestamp: 1469507970756,
info: {
method: "POST",
path: "/oauth/token",
headers: {
request: {
authorization: "Basic YWNtZTphY21lc2VjcmV0",
accept: "application/json, application/x-www-form-urlencoded",
content-type: "application/x-www-form-urlencoded",
cache-control: "no-cache",
pragma: "no-cache",
user-agent: "Java/1.8.0_65",
host: "localhost:8080",
connection: "keep-alive",
content-length: "101"
},
response: {
X-Content-Type-Options: "nosniff",
X-XSS-Protection: "1; mode=block",
Cache-Control: "no-cache, no-store, max-age=0, must-revalidate",
Pragma: "no-cache",
Expires: "0",
X-Frame-Options: "DENY",
X-Application-Context: "application",
Content-Type: "application/json;charset=UTF-8",
Transfer-Encoding: "chunked",
Date: "Tue, 26 Jul 2016 04:39:30 GMT",
status: "200"
}
}
}
},
{
timestamp: 1469507970664,
info: {
method: "GET",
path: "/oauth/authorize",
headers: {
request: {
accept: "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
accept-encoding: "gzip, deflate, sdch",
accept-language: "en-US,en;q=0.8",
connection: "keep-alive",
host: "localhost:8080",
upgrade-insecure-requests: "1",
user-agent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36"
},
response: {
X-Content-Type-Options: "nosniff",
X-XSS-Protection: "1; mode=block",
Cache-Control: "no-cache, no-store, max-age=0, must-revalidate",
Pragma: "no-cache",
Expires: "0",
X-Frame-Options: "DENY",
X-Application-Context: "application",
Location: "http://127.0.0.1:9999/client/login?code=bCG6Se&state=Bdn883",
status: "302"
}
}
}
},`

I failed to acknowledge that behind the scene there are these 3 requests for the oauth flow.
I see the following in the social application log:
o.s.s.w.a.i.FilterSecurityInterceptor    : Previously Authenticated: org.springframework.security.oauth2.provider.OAuth2Authentication

What I am trying to do incorrectly is: curl -v http://127.0.0.1:9999/client -H "authorization: bearer 9c03d406-14ac-4f98-a261-6ad29f81fe6a"
and I get 
`Connected to 127.0.0.1 (127.0.0.1) port 9999 (#0)
> GET /client HTTP/1.1
> User-Agent: curl/7.30.0
> Host: 127.0.0.1:9999
> Accept: */*
> authorization: bearer 9c03d406-14ac-4f98-a261-6ad29f81fe6a`

My expectation from step above was incorrect as I was expecting a response while it begins the OAuth workflow again. 
You are correct as request from either the browser or curl has exact same behavior.
Reading through https://spring.io/guides/tutorials/spring-boot-oauth2/#_social_login_authserver you mention:
_Password grant is appropriate for a native or mobile application, and where you have a local user database to store and validate the credentials. For a web app, or any app with "social" login, like ours, you need the "authorization code" grant, and that means you need a browser to handle redirects and render the user interfaces from the external providers._

I am conflating web server flow with user-name password flow. I did test my hypothesis here:
https://developers.google.com/oauthplayground
though I think google encodes a lot more in their OAuth token and hence a curl with just "bearer" token works once I grab the access token from their browser tool. I was wishfully thinking that I could cheat the web server flow to get access tokens and would not need to use password grant, but looks like thats the standard for my use case of independent REST requests. 
Thanks again for your attention, I am still trying to understand the concept and the implementation

**Comment by dsyer** [comment]
Good, so we are converging I think. I'm still not sure why you want to use curl, but at least you know how it works and where to get an access token now. I updated that section of the tutorial on password grant (it really isn't very often that useful), so thanks for the feedback, and the patience.
